### PR TITLE
Adding section on controlled types and exception handling

### PR DESCRIPTION
**Pull request by gusthoff**

### CODE REVIEW

File: content/courses/advanced-ada/parts/resource_management/controlled_types.rst [hunk 1]
```diff
@@ -1548,13 +1548,422 @@ Completion
...
+procedure is never called for object :ada:`A`.
+
+
+Bounded errors of controlled types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Bounded errors are an important topic when talking about exception and
+controlled types. In general, if an exception is raised in the :ada:`Adjust` or
+:ada:`Finalize` procedure, this is considered a bounded error |mdash| in its
+turn a bounded error raises a :ada:`Program_Error` exception.
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/resource_management/controlled_types.rst:1739]: It raises the P_E exception only if the error is detected, I believe.

File: content/courses/advanced-ada/parts/resource_management/controlled_types.rst [hunk 2]
```diff
@@ -1548,13 +1548,422 @@ Completion
...
+Now, we're calling the :ada:`Finalize` procedure explicitly in the
+:ada:`Show_Finalize_Exception` procedure. As we know, due to the operation on
+:ada:`I` in the :ada:`Finalize` procedure, the :ada:`Constraint_Error`
+exception is raised in the procedure. Because we're handling this exception in
+the :ada:`Show_Finalize_Exception` procedure, we see the corresponding user
+message ("Constraint_Error is being handled...") at runtime.
+
+(Note that in the exception handling block, we're calling the
+:ada:`Reset_Counter` procedure. This prevents the :ada:`Constraint_Error` to be
+raised in the next call to :ada:`Finalize`.)
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/resource_management/controlled_types.rst:1921]: Maybe "This prevents the :ada:`Constraint_Error` to be raised in the next call to :ada:`Finalize`." 
should be 
"This prevents :ada:`Constraint_Error` from being raised in the next call to :ada:`Finalize`."

File: content/courses/advanced-ada/parts/resource_management/controlled_types.rst [hunk 3]
```diff
@@ -1548,13 +1548,422 @@ Completion
...
+
+       --  When A is about to get out of
+       --  scope:
+       --
+       --  Finalize (A);
+       --
+    end Show_Simple_Exception;
+
+In this example, we're forcing an overflow to happen in the
+:ada:`Show_Simple_Exception` by adding one to the integer variable :ada:`Cnt`,
```
[the remark this answers is not in the document]
Comment (gusthoff) [reply content/courses/advanced-ada/parts/resource_management/controlled_types.rst:1631]: Ouch! Yes, you're right, thanks for the careful review!

File: content/courses/advanced-ada/parts/resource_management/controlled_types.rst (hunk 1 above)
Comment (gusthoff) [reply content/courses/advanced-ada/parts/resource_management/controlled_types.rst:1739]: Good point, I've changed the phrase accordingly.

File: content/courses/advanced-ada/parts/resource_management/controlled_types.rst (hunk 2 above)
Comment (gusthoff) [reply content/courses/advanced-ada/parts/resource_management/controlled_types.rst:1921]: Yes, that sounds better.

**Review by pat-rogers — APPROVED** [review]

**Pull request merged**